luci-app-packet-capture: add tcpdump application

wjowsa opened the pull request:
In some troubleshooting situations (e.g. catching logs by unprivileged user) is useful to run network troubleshooting tools like tcpdump or iperf3 directly from the Luci interface. This PR introduces a Troubleshooting menu and a first tool - Packet Capture. 
Packet Capture application is a fronted for tcpdump. With this application, a user is able to configure the most important and commonly used options of tcpdump. The output can be lived streamed to the browser or saved to a file.
The application uses the new [uhttpd API](https://git.openwrt.org/?p=project/uhttpd.git;a=commit;h=11723570af9cb7bd87842e79c85ee99530be9902). The EventSource interface is used in order to handle server-sent events. However, in order to make the API work properly, there are two patches for uhttpd that have to be applied.
 * Because the EventSource does not allow to set custom header fields, the uhttpd has patched to support passing rpcd session id in the subscription url. [patch 1](https://patchwork.ozlabs.org/project/openwrt/patch/20201110080209.6368-1-wojciech.jowsa@gmail.com/)
* The fix for uhttpd crash.  [patch  2](https://patchwork.ozlabs.org/project/openwrt/patch/20201115091917.7798-1-wojciech.jowsa@gmail.com/)
I have tested this application on TP-Link WDR4300 and PCEngines APU3 using the latest openwrt master.

![packet_capture](https://user-images.githubusercontent.com/4362399/99194502-05b55880-2780-11eb-94ee-8facfc47051e.gif)




Signed-off-by: Wojciech Jowsa <wojciech.jowsa@gmail.com>

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+	});
+}
+
+window.onbeforeunload = stopTcpdump;
```
Please use `window.addEventListener('beforeunload', stopTcpdump);`

wjowsa replied:
Ok

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+
+function stopTcpdump() {
+	fs.exec("/usr/sbin/packet_capture_stop").then(function(replay) {
+		if(eventSource) { 
```
Please put a space between the keyword (`if`) and the left paren (`if (eventSource)`). Also please omit the curly braces for single-statement loops or conditionals:

```
if (eventSource)
    eventSource.close();
```

wjowsa replied:
Ok.

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+});
+
+function findButton(title) {
+	var buttons = document.querySelectorAll('button');
```
This will not work once `Download` is translated into other languages. I suggest to add a dedicated attribute to the button, e.g. `data-action="download"` which you can match. This would also simplify this function into `return document.querySelector('button[data-action="..."]');`

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+
+function addOutput() {
+	var tcpdumpOut = document.querySelectorAll('[id$="tcpdump_out"]')[0];
+	if(tcpdumpOut) {
```
Space after `if` and no curly braces:
```
if (tcpdumpOut)
    return;
```

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+	frameEl.firstElementChild.style.fontFamily = 'monospace';
+
+	var downloadBtn = document.querySelectorAll('[id$="download_file"]')[0];
+	if(downloadBtn) {
```
Space after `if` and no curly braces

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+		enctype: 'application/x-www-form-urlencoded'
+	}, E('input', { type: 'hidden', name: 'sessionid', value: rpc.getSessionID()},
+	   E('input', { type: 'hidden', name: 'path', value: "/tmp/capture.pcap"},
+	   E('input', { type: 'hidden', name: 'filename', value: hostName + "-" + Date.now() + ".pcap"}
```
Suggest to add `E('input', { type: 'hidden', name: 'mimetype', value: 'application/vnd.tcpdump.pcap'})`

wjowsa replied:
Ok

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+		method: 'post',
+		action: '/cgi-bin/cgi-download',
+		enctype: 'application/x-www-form-urlencoded'
+	}, E('input', { type: 'hidden', name: 'sessionid', value: rpc.getSessionID()},
```
The inputs here are improperly nested:
```javascript
E('input', {...},
  E('input', {...},
    E('input', {...},
      ...)))
```

Should be:

```javascript
E('input', {...}),
E('input', {...}),
E('input', {...}),
...
```

wjowsa replied:
If I change nesting then only the first attribute (sessionid) and e.g. path is not passed.

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+	var tasks = [];
+	tasks.push(callLuciProcessList().then(L.bind(function(list) {
+		for (var i = 0; i < list.length; i++) {
+			if(list[i].COMMAND.includes("/usr/sbin/packet_capture")) {
```
For this and the subsequent `if` conditions:
 - add space after `if`
 - omit curly braces for single-statement conditionals

jow- commented on applications/luci-app-packet-capture/htdocs/luci-static/resources/view/packet_capture/tcpdump.js:
```diff
@@ -0,0 +1,293 @@
+	var tasks = [];
+	tasks.push(callLuciProcessList().then(L.bind(function(list) {
+		for (var i = 0; i < list.length; i++) {
+			if(list[i].COMMAND.includes("/usr/sbin/packet_capture")) {
```
Hm, so you're doing the equivalent of `ps auxww ·| grep /usr/sbin/packet_capture` here? This does not look very robust imho. Can't we extend `/usr/sbin/packet_capture` to provide some kind of status subcommand? E.g. something like `/usr/sbin/packet_capture is_running` which either exits with `0` (yes) or `1` (no) - or which outputs some JSON status info to stdout.

wjowsa replied:
I think that we can check whether /var/run/packet_capure.pid file exists or not? It works fine together with the changes you proposed below.

jow- commented on applications/luci-app-packet-capture/root/usr/sbin/packet_capture:
```diff
@@ -0,0 +1,61 @@
+#!/usr/bin/env lua
```
I believe this and the subsequent helper executables or not meant to be directly usable on the shell as normal commands, so please move them into `/usr/libexec/` to not pollute `$PATH`. Also I think all these scripts should be merged into one single executable.

wjowsa replied:
I moved scripts to libexec. One executable is a good idea but if you agree then I would like to keep separate scripts for now. Merging scripts sh and lua into one robust script might require some additional refactoring here and there so if possible then I would like to let users try this application. I promise that I will merge it into one exec in the nearest feature. More, I am going to push iperf3 application that is similar to the packet capture application. Iperf3 application will have a single executable. After that, I will copy the same structure to the packet capture application.

jow- commented on applications/luci-app-packet-capture/root/usr/sbin/packet_capture_start:
```diff
@@ -0,0 +1,42 @@
+fi
+
+if  [ -n "$interface" ];then
+	args="$args  -i  $interface"
```
Check for `-e "/sys/class/net/$interface"` to prevent shell injection via crafted `interface` JSON field value

jow- commented on applications/luci-app-packet-capture/root/usr/sbin/packet_capture_start:
```diff
@@ -0,0 +1,42 @@
+	args="$args  -i  $interface"
+fi
+
+if [ -n "$packets" ];then
```
Use something like  `if echo "$packets" | grep -qE '^[0-9]+$'; then` to prevent shell injection via crafted `packets` JSON field value

jow- commented on applications/luci-app-packet-capture/root/usr/sbin/packet_capture_start:
```diff
@@ -0,0 +1,42 @@
+	args="$args -W 2 -C $((mem/(1024 * 10))) -w /tmp/capture.pcap -z /usr/sbin/packet_capture_stop"
+fi
+
+if [ -n "$filter" ];then
```
I suggest to use something like `if [ -n "$filter" ] && tcpdump -i lo -d "$filter" >/dev/null 2>/dev/null; then` to test the filter expression for validity.

jow- commented on applications/luci-app-packet-capture/root/usr/sbin/packet_capture_start:
```diff
@@ -0,0 +1,42 @@
+    echo "$filter" >> /tmp/tcpdump_filter
+fi
+
+(/usr/sbin/packet_capture "$args" "$duration")&
```
I suggest to write the resulting pid (`$!`) into a pid file...

jow- commented on applications/luci-app-packet-capture/root/usr/sbin/packet_capture_stop:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+pgrep tcpdump | xargs kill -9
```
... and use that pid file here to avoid killing unrelated processes.

```bash
pid=$(cat /var/run/packet_capture.pid)
if [ -n "$pid" ] && grep -sq tcpdump "/proc/$pid/cmdline"; then
    kill -TERM $pid
fi
rm -f /var/run/packet_capture.pid
```

wjowsa replied:
Actually tcpdump process has to be terminated to that is does not become zombie

jow- commented on applications/luci-app-packet-capture/root/usr/share/luci/menu.d/luci-app-packet-capture.json:
```diff
@@ -0,0 +1,25 @@
+			"path": "packet_capture/tcpdump"
+		},
+		"depends" : {
+            "acl": [ "luci-app-packet-capture" ],
```
The indentation here appears to be inconsistent

jow- commented on applications/luci-app-packet-capture/root/usr/share/rpcd/acl.d/luci-app-packet-capture.json:
```diff
@@ -0,0 +1,24 @@
+			},
+			"uci": [ "packet_capture", "system" ],
+			"file": {
+				"/tmp/capture.pcap": [ "read", "write" ]
```
Please grant only `read` permission here...

jow- commented on applications/luci-app-packet-capture/root/usr/share/rpcd/acl.d/luci-app-packet-capture.json:
```diff
@@ -0,0 +1,24 @@
+		},
+		"write": {
+			"uci": [ "packet_capture" ],
+			"file": {
```
... and `"/tmp/capture.pcap": [ "write" ]` here.

jow- commented:
Thanks a lot for your contribution @wjowsa - I added a couple of review comments.

wjowsa commented:
Thanks for the review @jow- , I will apply your comments soon.
The credit goes also to @kristrev who helped me with the design and development of this application.  I hope it will make debugging of networking issues more user-friendly :)

wjowsa commented:
@jow- I think I have addressed all your comments. Unfortunately, tcpdump does allow to set maximum capture file size yet so I had to use a workaround to add this functionality. However, there is a PR https://github.com/the-tcpdump-group/tcpdump/pull/781 which adds the file max-size option. It works fine so potentially it could be added to tcpdump package. Then we could remove the file max-size workaround.

systemcrash commented:
@wjowsa please address all comments, or you can close this PR. ( It might also help to drive PR you mentioned above for max sizes )
